### ISSUE TITLE
Selectable descriptor for pcap file stops being readable with kqueue

### ISSUE THREAD

**Issue by timwoj** [the-tcpdump-group/libpcap#870]
libpcap version: `1.8.1 -- Apple version 89.11.2`
OS name/version: macOS 15.0.1
Configure flags: unknown (macOS defaults?)

I'm running into an issue attempting to read a pcap file using `kqueue`, where it gets a certain number of packets into the file and then gets "stuck". `kevent` doesn't ever return that the file descriptor (from `pcap_get_selectable_fd`) is ready for reading, and my test code just sits waiting forever.

A simple gist with the test code in it: https://gist.github.com/timwoj/1efbd376c27cfd217751ccc6f2cbc58f
The sample file I've been testing with: https://github.com/zeek/zeek/blob/master/testing/btest/Traces/http/get.trace

If this would be a better question for the mailing list, let me know and I can repost it over there.

**Comment by timwoj** [comment]
I retested it with version 1.9.1 from homebrew and it has the same result of getting stuck at a certain point in the file.

**Comment by guyharris** [comment]
To quote the 10.14.6 KQUEUE(2) man page:

     EVFILT_READ      Takes a file descriptor as the identifier, and returns
		      whenever there is data available to read.  The behavior
		      of the filter is slightly different depending on the
		      descriptor type.

			...

		      Vnodes
			  Returns when the file pointer is not at the end of
			  file.  data contains the offset from current posi-
			  tion to end of file, and may be negative.

If the pointer *is* at the end of the file, it doesn't return; the intent is that it be used when doing a "tail -f"-style operation (tail, in fact, uses it to implement -f).

pcap uses the standard I/O library routines to read captures (and that's baked into the API, given `pcap_fopen_offline()` and `pcap_file()`).  That library is buffered, so a single `read()` (or, on Windows, `ReadFile()`) done by the library may read more than one packet.  This means that the file descriptor might be at the end of file even if you haven't read all the packets - there may be additional unread packets in the buffer.

So you can't really use the selectable FD (which, for a capture file, is just the result of `fileno()` on the `FILE *` for the file) to do a "tail -f"-style operation.  It also may not work well if you're trying to read from a pipe in non-blocking mode.

It might be possible to support non-blocking mode for `pcap_t`s from `pcap_open_offline()` in ways that would allow those operations; it would also require having `pcap_dispatch()` to return more than one packet from such a `pcap_t`.

It's probably not a trivial change.

**Comment by timwoj** [comment]
Dang, I was hoping for a way to check whether a pcap file has packets left to process without having to call `pcap_dispatch` or `pcap_next_ex` and inspecting the return value. I had some other file descriptors I'm already polling on in a loop, so having a working file descriptor here would be best.

**Comment by guyharris** [comment]
"Has packets left to process" as in "I'm reading a file that's not being in the process of being written, so I want to see if I've hit the end of the file" or "I'm doing a tail -f-style read of the file, so I want to see if more packets have been added"?

If it's the former, what else are you doing in the loop in parallel with reading the file?

**Comment by timwoj** [comment]
I work on the [Zeek](http://zeek.org) project. We use libpcap for reading trace files and network interfaces for packet data. The current implementation effectively uses a busy loop to spin on the pcap object looking for new data, which admittedly uses more CPU time than it should.

We're investigating ways that we can reimplement the main loop, including hopefully changing to poll the pcap object directly for data. We also poll against some sockets for things like DNS requests and an IPC mechanism called broker. It'd be great if we could just add the pcap interface into the same poll.

**Comment by guyharris** [comment]
> I work on the [Zeek](http://zeek.org) project. We use libpcap for reading trace files and network interfaces for packet data. The current implementation effectively uses a busy loop to spin on the pcap object looking for new data, which admittedly uses more CPU time than it should.

libpcap arguably made a mistake using the same data type for live capture handles and capture file handles; for a lot of purposes, the capture loop and read loop are different.

> We're investigating ways that we can reimplement the main loop, including hopefully changing to poll the pcap object directly for data.

I'm not sure the concept of a "pcap object" is useful, as per the above.

If the main loop is reading from a capture file, is it reading from a file to which some other program is writing, so that if it reaches the current end-of-file, that doesn't mean it's done, it means it should wait for more data to arrive?  Or is it always reading from a complete file?

If it's always reading from a complete file, there's never "new data" - all the data you have is there from the start, and the main loop should reflect that, as there's nothing for which to poll.

If it's reading from a file that's being written to by another program (a "tail -f"-style read), that will require some work on your part and possibly work on our part as well.

**Comment by timwoj** [comment]
> If the main loop is reading from a capture file, is it reading from a file to which some other program is writing, so that if it reaches the current end-of-file, that doesn't mean it's done, it means it should wait for more data to arrive? Or is it always reading from a complete file?

It's always reading from a complete file.

> If it's always reading from a complete file, there's never "new data" - all the data you have is there from the start, and the main loop should reflect that, as there's nothing for which to poll.

This is true. We're just trying to maintain that the "live" and "capture" versions of the loop look roughly the same. You wait for data to come in, and you process it. In the end, perhaps we should just set the poll timeout to zero in the "capture" case and let it spin on the data.

That said, does the file descriptor work as expected if reading from a network interface with live data?

**Comment by guyharris** [comment]
> We're just trying to maintain that the "live" and "capture" versions of the loop look roughly the same.

Unfortunately, kqueues treat descriptors for regular files in a fashion that's convenient for "tail -f" but is inconvenient for code that wants to treat them the same way as other file descriptors, as you've found.  This makes treating them the same difficult if you're going to use a kqueue-based loop.  A `select()`-based or `poll()`-based loop, where the descriptor is, as far as I know, *always* considered readable, even if a read gets an EOF, could work that way, but a queue-based loop can't.

> You wait for data to come in, and you process it.

For a file, that's "you open the file, you call `pcap_loop()`, you're done".  The data all comes in the instant you read the file, so you process it all in one gulp.

> That said, does the file descriptor work as expected if reading from a network interface with live data?

It should do so, modulo the warnings in the pcap_get_selectable_fd man page.  Note that sometimes `pcap_get_selectable_fd()` can return -1 for a live capture device, meaning "there is no selectable FD".  We currently do that for a few cases on systems using BPF where we know that particular version of the OS won't even let select/poll work if you put the `pcap_t` into non-blocking mode, add a timeout to the select/poll, and try to read the descriptor if the timeout expires (FreeBSD 4.3 and  4.4).

We also do that for Endace DAG devices; they don't have an FD *at all*, and don't block in the driver waiting for packets to arrive - they just poll the device.  To handle *that*, newer version of libpcap have `pcap_get_required_select_timeout()`, which returns a timeout that should be used in select/poll/whatever - the device should be put into non-blocking mode and a `pcap_dispatch()` should be done if the timeout expires.

**Comment by timwoj** [comment]
Thanks for all of the info. I think we've got a good implementation for packet files now and everything is working.

For live interfaces though, I ran into a weird problem that only seems to affect `kqueue` and not `poll` or `epoll`. I've only tested on macOS with the default libpcap (1.8.0). I haven't tested on one of the other BSDs yet to see if it's macOS-specific.

The problem is that `kevent()` doesn't always see the interface as ready. Sometimes it will just sit and not read any packets at all. This happens whether the file descriptor for libpcap is the only thing in the queue or not. On rare occasion, everything works perfectly and we get packets from ilbpcap. This is on a known-live interface with a ping running every second in the background just to make sure there's traffic flowing. Is there any easy way to debug what's happening here?

**Comment by mcr** [comment]
It's unclear to me if there is anything that can really change.  It seems like maybe we just need a BCP on using kqueue?

**Comment by guyharris** [comment]
> For live interfaces though, I ran into a weird problem that only seems to affect `kqueue` and not `poll` or `epoll`. I've only tested on macOS with the default libpcap (1.8.0). I haven't tested on one of the other BSDs yet to see if it's macOS-specific.
> 
> The problem is that `kevent()` doesn't always see the interface as ready. Sometimes it will just sit and not read any packets at all. This happens whether the file descriptor for libpcap is the only thing in the queue or not. On rare occasion, everything works perfectly and we get packets from ilbpcap. This is on a known-live interface with a ping running every second in the background just to make sure there's traffic flowing. Is there any easy way to debug what's happening here?

That's a separate issue; try it on FreeBSD (the macOS implementation of "make the timeout work with `select()`/`poll()`/kqueues" was based on FreeBSD's, with necessary tweaks made to cope with different kernel timer APIs, and they both share at least one glitch where doing an ioctl on the BPF descriptor in another thread stops the timer on which that depends, so they may share other glitches as well).

**Comment by guyharris** [comment]
> It's unclear to me if there is anything that can really change. It seems like maybe we just need a BCP on using kqueue?

Yes, as per my comment, you have to be careful when you use kqueues with regular files.

**Comment by infrastation** [comment]
Does this discussion boil down to anything that should be copied to the documentation? If not, this issue looks resolved and will be closed in 7 days.

**Comment by timwoj** [comment]
I'm fine with closing it now. There's nothing to do code-wise.

**Comment by infrastation** [comment]
Alright, closing then.
